packages/eslint-plugin/docs/rules/no-wrapper-object-types.mdx

```diff
@@ -0,0 +1,76 @@
+---
+description: 'Disallow using confusing built-in primitive class wrappers.'
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+> 🛑 This file is source code, not the primary documentation location! 🛑
+>
+> See **https://typescript-eslint.io/rules/no-wrapper-object-types** for documentation.
+
+The JavaScript language has a set of language types, but some of them correspond to two TypeScript types, which look similar: `boolean`/`Boolean`, `number`/`Number`, `string`/`String`, `bigint`/`BigInt`, `symbol`/`Symbol`, `object`/`Object`.
+The difference is that the lowercase variants are compiler intrinsics and specify the actual _runtime types_ (that is, the return value when you use the `typeof` operator), while the uppercase variants are _structural types_ defined in the library that can be satisfied by any user-defined object with the right properties, not just the real primitives.
+JavaScript also has a "wrapper" class object for each of those primitives, such as [`Boolean`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean) and [`Number`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number).
+These wrapper objects are assignable to the uppercase types, but not to the lowercase types.
+
+Using the primitives like `0` instead of object wrappers like `new Number(0)` is generally considered a JavaScript best practice.
+JavaScript programs typically work with the real number primitives, rather than objects that "look like" numbers.
+As a result, using the lowercase type names like `number` instead of the uppercase names like `Number` helps make your code behave more reliably.
+
+TypeScript's `Function` type suffers from the same issue of structural typing.
+In addition, `Function` allows being called with any number of arguments and returns type `any`, while in reality, it may be a class which is never callable, or it may be a plain object that happens to possess all properties of the `Function` class!
+It's generally better to specify function parameters and return types with the function type syntax.
```

Comment:
@Josh-Cena btw I removed this from the suggestion:

> `If you want a generic "callable" type, use `(...args: never) => unknown`.`

I'm really scared of giving people a single syntax reference. The generic "callable" type one would want changes based on the scenario. And since this is early on in the doc, folks are likely to mistreat it as a true canonical reference.

Reply:
Hmmm, ok. No strong opinions but still think it would be helpful to have some words about alternatives.

Reply:
One advantage of having a dedicated rule (https://github.com/typescript-eslint/typescript-eslint/pull/9102#issuecomment-2142874497 -> https://github.com/typescript-eslint/typescript-eslint/pull/9102#issuecomment-2143965277) is we can get a bit more dedicated space in the docs. So, what are the different function forms we might want to mention?

Maybe:

* `() => void`: a function that has no parameters and whose return is ignored
* `(...args: never) => unknown`: a general "callable" type
* `(...args: unknown[]) => unknown`: a general "callable" type, also?

...I personally don't know of general practice for choosing between the `: never` and the `unknown[]`.

cc @bradzacher from https://github.com/typescript-eslint/typescript-eslint/pull/9102#discussion_r1616432707

Reply:
`(...args: never) => unknown` is the top type for callables. Every function is assignable to it but itself is not callable because you don't know what to call it with. `(...args: unknown[]) => unknown` is actually callable with anything, but only things that accept `...args: unknown[]` (like `console.log`) are assignable to it, so it's pretty useless in practice.

I think when people use `Function` to mean "any function" they mean `(...args: never) => unknown`. It soundly forbids you from calling it.

Reply:
Thanks, that's a great explanation 😄. 2508d846f